File: javascript/mixed-juices/mixed-juices.js
// @ts-check
//
// The line above enables type checking for this file. Various IDEs interpret
// the @ts-check directive. It will give you helpful autocompletion when
// implementing this exercise.

/**
 * Determines how long it takes to prepare a certain juice.
 *
 * @param {string} name
 * @returns {number} time in minutes
 */
export function timeToMixJuice(name) {
  switch (name) {
    case 'Pure Strawberry Joy':
      return 0.5;
    case 'Energizer':
    case 'Green Garden':
      return 1.5;
    case 'Tropical Island':
      return 3;
    case 'All or Nothing':
      return 5
    default:
      return 2.5;
  };
}

const WEDGES_PER_LIME = {
  "small": 6,
  "medium": 8,
  "large": 10
};

/**
 * Calculates the number of limes that need to be cut
 * to reach a certain supply.
 *
 * @param {number} wedgesNeeded
 * @param {string[]} limes
 * @returns {number} number of limes cut
 */
export function limesToCut(wedgesNeeded, limes) {
  let sum = 0, i = 0;
  while (sum < wedgesNeeded && i < limes.length) {
    sum += WEDGES_PER_LIME[limes[i]];
    i++;
  }
  return i;
}

/**
 * Determines which juices still need to be prepared after the end of the shift.
 *
 * @param {number} timeLeft
 * @param {string[]} orders
 * @returns {string[]} remaining orders after the time is up
 */
export function remainingOrders(timeLeft, orders) {
  let i = 0;
  for (; timeLeft > 0 && i < orders.length; i++)
    timeLeft -= timeToMixJuice(orders[i]);
  return orders.slice(i);
}

File: julia/mixed-juices/mixed-juices.jl
const mix_times = Dict(
    "Pure Strawberry Joy" => 0.5,
    "Energizer" => 1.5,
    "Green Garden" => 1.5,
    "Tropical Island" => 3.0,
    "All or Nothing" => 5.0
)
const default_mix_time = 2.5
    
time_to_mix_juice(juice) = get(mix_times, juice, default_mix_time)
wedges_from_lime(size)   = get(Dict("small" => 6, "medium" => 8, "large" => 10), size, 0)

function limes_to_cut(needed, limes)
    i = 0
    while needed > 0 && i < length(limes)
        i += 1
        needed -= wedges_from_lime(limes[i])
    end
    return i
end

# order_times(orders) = map(time_to_mix_juice, orders)
# or, practice this concept
function order_times(orders)
    times = []
    for juice in orders
        push!(times, time_to_mix_juice(juice))
    end
    times
end

function remaining_orders(time_left, orders)
    while time_left > 0 && !isempty(orders)
        time_left -= time_to_mix_juice(popfirst!(orders))
    end
    orders
end
